### ISSUE TITLE
Potential mistakes in the code

### ISSUE THREAD

**Issue by WU-YU-TONG** [Harvard-Ophthalmology-AI-Lab/FairCLIP#8]
Congrats a lot to your contributions. 

I discovered a mistake in the code. In ./FairCLIP/finetune_FairCLIP.py, line 190, the calculated correlation is derived directly from the logits, which is not normalized. Yet in line 202 the correlation by groups is normalized. As below:
```
#line 190
correlations_with_batch = similarity.diag().float() # not normalized
correlations_groups = []

for x in group_dataloaders:
    images_dist, texts_dist, label_and_attributes_dist = next(x)
    images_dist= images_dist.to(device)
    texts_dist = texts_dist.to(device)
    with torch.no_grad():
        img_feats, txt_feats = model(images_dist, texts_dist)
    similarity = (img_feats @ txt_feats.T)
    correlations_with_group = similarity.diag().float()
    correlations_with_group /= correlations_with_group.sum() # normalized here
    total_loss = total_loss + args.lambda_fairloss * loss_for_FairCLIP(correlations_with_batch[:,None], correlations_with_group[:,None])
   #normalized and unormalized correlations are treated equally here.
```
Should there be a normalization operation right after line 190?

In my perspective, this mistake makes the loss_for_FairCLIP totally meaningless, or even harmful to the overall performance, as the correlation before normalization can be hundreds times bigger than the normalized one.

**Comment by luoyan407** [comment]
Thank you for your interest in our project and for your thoughtful question.

`correlations_with_batch` is derived from unnormalized logits, while `correlations_with_group` is explicitly normalized to sum to one. However, this does not necessarily mean the loss is meaningless or harmful. Instead, this asymmetry can be viewed as a regularization technique, and it is a plausible and coherent strategy.

The reasons are three-fold as follows.

1. The main contrastive loss (total_loss on line 191) works by pushing the similarity scores of correct image-text pairs higher. In contrast, the `loss_for_FairCLIP` term, by comparing large-valued unnormalized scores to small-valued normalized scores, penalizes high correlation values. The training process, therefore, becomes a balancing act between the primary loss increasing the scores and this fairness term regularizing them by pushing them down.

2. The `lambda_fairloss` coefficient is very small (e.g., 1e-4). This ensures the regularizing effect is a gentle pressure, not a dominant force that would nullify the primary learning signal. It is meant to moderate the scores, preventing the model from becoming overly confident in a way that might exacerbate biases, rather than causing "harm" by crushing the scores entirely.

3. The code iterates through dataloaders for different demographic groups and adds a `loss_for_FairCLIP` term for each one. This means the model is encouraged to produce correlation scores that are not just similar in distribution shape across groups, but also similar in scale. By penalizing the magnitude of the main batch's correlations relative to a consistent, normalized target for each group, the model is discouraged from systematically producing higher scores for one group over another.
